include/pybind11/pybind11.h

```diff
@@ -1954,25 +1954,49 @@ inline std::pair<decltype(internals::registered_types_py)::iterator, bool> all_t
     return res;
 }
 
-template <typename Iterator, typename Sentinel, bool KeyIterator, return_value_policy Policy>
+template <typename Access, return_value_policy Policy, typename Iterator, typename Sentinel, typename ValueType, typename... Extra>
 struct iterator_state {
     Iterator it;
     Sentinel end;
     bool first_or_done;
 };
 
-PYBIND11_NAMESPACE_END(detail)
+// Note: these helpers take the iterator by non-const reference because some
+// iterators in the wild can't be dereferenced when const.
+template <typename Iterator>
+struct iterator_access {
+    using result_type = decltype((*std::declval<Iterator>()));
+    // NOLINTNEXTLINE(readability-const-return-type) // PR #3263
+    result_type operator()(Iterator &it) const {
```

Comment:
Does the compiler automatically inline these short functions? If not, shall we add `inline` here?

There may be some size reduction if the compiler does not need to generating symbols for each template instantiation.

Reply:
Member functions defined inside the class implicitly have inline linkage. That doesn't mean the compiler *has* to inline them, but since the binary size is identical to master (when using the same test suite) I think it's safe to assume they're inlined.